pgxn/neon/neon.c

```diff
@@ -266,6 +267,175 @@ LogicalSlotsMonitorMain(Datum main_arg)
+	 * PGPROC_MAX_CACHED_SUBXIDS. But it is too small. Also we take in account
+	 * only subxacts, not top level transactions which number can be grreater
+	 * thsn 64.
+	 *
+	 * Summarizing all above: there are following apporaches and related
+	 * risks: 1. If we restore as much running xacts as it possible
+	 * (TOTAL_MAX_CACHED_SUBXIDS) then there is risk that future transactions
+	 * will trigger e "too many KnownAssignedXids" error which cause crash of
+	 * replica. 2. If we limit number of restored xacts by some small constant
+	 * (PGPROC_MAX_CACHED_SUBXIDS), then miost likely we will have to wait
+	 * until primary send us running-xacts non-overflowed WAL record. It make
+	 * takes quite long time (running xacts records is produced erach 15
+	 * seconds) and if primary is offline it will never happen and sop replica
+	 * is not started. 3. If we restore just some subset of subxacts but
+	 * doesn;t mark snapshot as overflown (and so do not wait for primary),
+	 * then there is a risk of data corruption (wrong hint bits) or incorrect
+	 * query results (brtoken visibility) at replica.
+	 *
+	 * Current solution is to limit subxacts by  TOTAL_MAX_CACHED_SUBXIDS / 2
+	 * and mark snmapshot as overflowed if real number if larger.
+	 */
+	max_xcnt = TOTAL_MAX_CACHED_SUBXIDS / 2;
+
+	if (!TransactionIdIsNormal(from))
+	{
+		/*
+		 * No checkpoint or running-xacts record was written, so we do not
+		 * know oldestActiveXid. So we do not have low boundary for XIDs and
+		 * stop recovery after processing up to max_xcnt transactions.
+		 */
+		xcnt = max_xcnt;
+	}
+	else
+	{
+		xcnt = till > from ? till - from : till - from - FirstNormalTransactionId;
+		xcnt = Min(xcnt, max_xcnt);
+	}
+	*xids = (TransactionId *) palloc(Min(xcnt, max_xcnt) * sizeof(TransactionId));
+	xcnt = 0;
+	ConditionVariableCancelSleep();
+	PG_TRY();
+	{
+		do
+		{
+			TransactionIdRetreat(xid);
+			if (xid == till)
+			{
+				/*
+				 * Unlikely case: if low boundary (oldestActiveXid) is
+				 * unknown, and CLOG is not truncated, we can traverse all
+				 * 2^32 XIDs. To preventinfinite loop check if we reach upper
+				 * boundary.
+				 */
+				break;
+			}
+			xidstatus = TransactionIdGetStatus(xid, &xidlsn);
+
+			if (xidstatus == TRANSACTION_STATUS_IN_PROGRESS)
+			{
+				if (xcnt < max_xcnt)
+				{
+					(*xids)[xcnt] = xid;
+				}
+				else
+				{
+					/*
+					 * Overflowed. We won't be able to install the
+					 * RunningTransactions snapshot. We could bail out now,
+					 * but keep going to count the XIDs, for the sake of the
+					 * LOG message at the end of the function.
+					 */
+					overflow = true;
+					break;
+				}
+				xcnt += 1;
+			}
+		} while (xid != from);
+	}
+	PG_CATCH();
+	{
+		/*
+		 * CLOG can be already truncated, so be ready to catch
+		 * TransactionIdGetStatus error
+		 */
+		elog(LOG, "RestoreRunningXactsFromClog[%d..%d]: failed to get status of transaction %d",
+			 from, till, xid);
```

Comment:
Seems like a dangerous assumption. If there's an error reading the CLOG segment from pageserver, for example, we'll just ignore it and continue with the startup.

Can you write a test case for this, please?

Reply:
I removed TRY/CATCH clause.
Do we need to add some more test here?